### PR TITLE
Migrate pallet-proxy to pallet attribute macro

### PR DESCRIPTION
**Pull request by emostov**
Part of #7882.

Converts the `Proxy` pallet to the new pallet attribute macro introduced in #6877.

[Upgrade guidelines used](https://substrate.dev/rustdocs/v3.0.0/frame_support/attr.pallet.html#upgrade-guidelines).

## ⚠️ Breaking Change  ⚠️ 

From [checking upgrade guidelines](https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines)

> storages now use PalletInfo for module_prefix instead of the one given to `decl_storage`: use of this pallet in `construct_runtime!` needs careful updating of the name in order to not break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Proxy` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Proxy` pallet.

### Notes

There are some changes to the docs in metadata for the constants. The docs in the metadata for constants are now more complete.

### CODE REVIEW

**Comment by emostov** [comment]
Metadata diff

```bash
$ diff -c post-proxy-meta.json pre-proxy-meta.json
*** post-proxy-meta.json        2021-03-15 19:52:29.000000000 -0700
--- pre-proxy-meta.json 2021-03-14 19:17:22.000000000 -0700
***************
*** 395,401 ****
          }
        ],
        "documentation": [
!         " Dispatch the given `call` from an account that the sender is authorized for through",
          " `add_proxy`.",
          "",
          " Removes any corresponding announcement(s).",
--- 395,401 ----
          }
        ],
        "documentation": [
!         " Dispatch the given `call` from an account that the sender is authorised for through",
          " `add_proxy`.",
          "",
          " Removes any corresponding announcement(s).",
***************
*** 473,482 ****
          0
        ],
        "documentation": [
!         " The base amount of currency needed to reserve for creating a proxy.",
!         "",
!         " This is held for an additional storage item whose value size is",
!         " `sizeof(Balance)` bytes and whose key size is `sizeof(AccountId)` bytes."
        ]
      },
      {
--- 473,479 ----
          0
        ],
        "documentation": [
!         " The base amount of currency needed to reserve for creating a proxy."
        ]
      },
      {
***************
*** 501,510 ****
          0
        ],
        "documentation": [
!         " The amount of currency needed per proxy added.",
!         "",
!         " This is held for adding 32 bytes plus an instance of `ProxyType` more into a pre-existing",
!         " storage value."
        ]
      },
      {
--- 498,504 ----
          0
        ],
        "documentation": [
!         " The amount of currency needed per proxy added."
        ]
      },
      {
***************
*** 528,534 ****
          0
        ],
        "documentation": [
!         " The maximum amount of time-delayed announcements that are allowed to be pending."
        ]
      },
      {
--- 522,528 ----
          0
        ],
        "documentation": [
!         " `MaxPending` metadata shadow."
        ]
      },
      {
***************
*** 553,561 ****
          0
        ],
        "documentation": [
!         " The base amount of currency needed to reserve for creating an announcement.",
!         "",
!         " This is held when a new storage item holding a `Balance` is created (typically 16 bytes)."
        ]
      },
      {
--- 547,553 ----
          0
        ],
        "documentation": [
!         " `AnnouncementDepositBase` metadata shadow."
        ]
      },
      {
***************
*** 580,589 ****
          0
        ],
        "documentation": [
!         " The amount of currency needed per announcement made.",
!         "",
!         " This is held for adding an `AccountId`, `Hash` and `BlockNumber` (typically 68 bytes)",
!         " into a pre-existing storage value."
        ]
      }
    ],
--- 572,578 ----
          0
        ],
        "documentation": [
!         " `AnnouncementDepositFactor` metadata shadow."
        ]
      }
    ],
```

File: frame/proxy/src/lib.rs [hunk 1]
```diff
@@ -137,84 +89,84 @@ pub struct Announcement<AccountId, Hash, BlockNumber> {
...
-	trait Store for Module<T: Config> as Proxy {
-		/// The set of account proxies. Maps the account which has delegated to the accounts
-		/// which are being delegated to, together with the amount held on deposit.
-		pub Proxies get(fn proxies): map hasher(twox_64_concat) T::AccountId
-			=> (Vec<ProxyDefinition<T::AccountId, T::ProxyType, T::BlockNumber>>, BalanceOf<T>);
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::{ *, DispatchResult };
```
Comment (thiolliere) [inline frame/proxy/src/lib.rs:96]:
```suggestion
	use super::{*, DispatchResult};
```
Interesting, rust cannot disambiguate the same type.

File: frame/proxy/src/lib.rs [hunk 2]
```diff
@@ -523,8 +515,72 @@ decl_module! {
...
+	#[pallet::getter(fn proxies)]
+	pub type Proxies<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		(Vec<ProxyDefinition<T::AccountId, T::ProxyType, T::BlockNumber>>, BalanceOf<T>),
+		ValueQuery
+	>;
+
+    /// The announcements made by the proxy (key).
```
Comment (thiolliere) [inline frame/proxy/src/lib.rs:573]:
```suggestion
	/// The announcements made by the proxy (key).
```

**Review by thiolliere — APPROVED** [review]

File: frame/proxy/src/lib.rs [hunk 3]
```diff
@@ -137,84 +89,84 @@ pub struct Announcement<AccountId, Hash, BlockNumber> {
...
+		/// This is held for an additional storage item whose value size is
+		/// `sizeof(Balance)` bytes and whose key size is `sizeof(AccountId)` bytes.
+		#[pallet::constant]
+		type ProxyDepositBase: Get<BalanceOf<Self>>;
 
 		/// The amount of currency needed per proxy added.
-		const ProxyDepositFactor: BalanceOf<T> = T::ProxyDepositFactor::get();
+		///
+		/// This is held for adding 32 bytes plus an instance of `ProxyType` more into a pre-existing
+		/// storage value.
```
Comment (dvdplm) [inline frame/proxy/src/lib.rs:133]: I don't understand what this doc string says; can you rephrase it a bit you think?

File: frame/proxy/src/lib.rs [hunk 4]
```diff
@@ -137,84 +89,84 @@ pub struct Announcement<AccountId, Hash, BlockNumber> {
...
+		/// The maximum amount of time-delayed announcements that are allowed to be pending.
+		#[pallet::constant]
+		type MaxPending: Get<u32>;
 
-		/// `AnnouncementDepositBase` metadata shadow.
-		const AnnouncementDepositBase: BalanceOf<T> = T::AnnouncementDepositBase::get();
+		/// The type of hash used for hashing the call.
+		type CallHasher: Hash;
+
+		/// The base amount of currency needed to reserve for creating an announcement.
```
Comment (dvdplm) [inline frame/proxy/src/lib.rs:151]:
```suggestion
		/// The base amount of currency needed in reserve for creating an announcement.
```

File: frame/proxy/src/lib.rs [hunk 5]
```diff
@@ -476,13 +459,21 @@ decl_module! {
...
+			call_hash: CallHashOf<T>
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 			Self::edit_announcements(&delegate, |ann| ann.real != who || ann.call_hash != call_hash)?;
+
+			Ok(().into())
 		}
 
-		/// Dispatch the given `call` from an account that the sender is authorised for through
+		/// Dispatch the given `call` from an account that the sender is authorized for through
```
Comment (dvdplm) [inline frame/proxy/src/lib.rs:476]: I think we have a preference for british spelling. ;)

**Review by dvdplm — APPROVED** [review]
lgtm, modulo doc nitpicks.

File: frame/proxy/src/lib.rs (hunk 4 above)
Comment (emostov) [reply frame/proxy/src/lib.rs:151]: Correct me if you think I am wrong but I think the current version makes more sense. When you make an announcement you will need "to" reserve at least this amount (so it would move from free => reserve balance). But the balance is not already "in" reserve prior to making the announcement.

File: frame/proxy/src/lib.rs (hunk 5 above)
Comment (emostov) [reply frame/proxy/src/lib.rs:476]: TIL

File: frame/proxy/src/lib.rs (hunk 4 above)
Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:151]: yes i agree

File: frame/proxy/src/lib.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:133]: it is saying that we should modify `ProxyDepositFactor` to take into account 32 bytes of data + `proxytype.encode().len()` more bytes of data.

It is a hint for the person configuring this pallet to pick a reasonable number for a deposit.

File: frame/proxy/src/lib.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:476]: def no, we should stick with US spelling where possible, as this is standard for Computer Science.

Any UK spelling should ideally be corrected.

Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:476]: see https://github.com/paritytech/substrate/pull/4940

File: frame/proxy/src/lib.rs (hunk 3 above)
Comment (emostov) [reply frame/proxy/src/lib.rs:133]: @shawntabrizi do you think it is worth updating the comment or should it already be clear enough to the target audience?

Heres a sample of how it could be updated, but not sure it really adds much.
```rust
/// This is held for adding 32 bytes plus an instance of `ProxyType` more into a pre-existing
/// storage value. Thus, when configuring `ProxyDepositFactor` one should take into account 
/// `32 + proxytype.encode().len()` bytes of data.
```

Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:133]: yeah that looks like a reasonable addition :)

Couldn't hurt i think.
